File: wpilibj/src/main/java/edu/wpi/first/wpilibj/DMA.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj;

import edu.wpi.first.hal.DMAJNI;
import edu.wpi.first.wpilibj.motorcontrol.PWMMotorController;

/** Class for configuring Direct Memory Access (DMA) of FPGA inputs. */
public class DMA implements AutoCloseable {
  final int m_dmaHandle;

  /** Default constructor. */
  public DMA() {
    m_dmaHandle = DMAJNI.initialize();
  }

  @Override
  public void close() {
    DMAJNI.free(m_dmaHandle);
  }

  /**
   * Sets whether DMA is paused.
   *
   * @param pause True pauses DMA.
   */
  public void setPause(boolean pause) {
    DMAJNI.setPause(m_dmaHandle, pause);
  }

  /**
   * Sets DMA to trigger at an interval.
   *
   * @param periodSeconds Period at which to trigger DMA in seconds.
   */
  public void setTimedTrigger(double periodSeconds) {
    DMAJNI.setTimedTrigger(m_dmaHandle, periodSeconds);
  }

  /**
   * Sets number of DMA cycles to trigger.
   *
   * @param cycles Number of cycles.
   */
  public void setTimedTriggerCycles(int cycles) {
    DMAJNI.setTimedTriggerCycles(m_dmaHandle, cycles);
  }

  /**
   * Adds position data for an encoder to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param encoder Encoder to add to DMA.
   */
  public void addEncoder(Encoder encoder) {
    DMAJNI.addEncoder(m_dmaHandle, encoder.m_encoder);
  }

  /**
   * Adds timer data for an encoder to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param encoder Encoder to add to DMA.
   */
  public void addEncoderPeriod(Encoder encoder) {
    DMAJNI.addEncoderPeriod(m_dmaHandle, encoder.m_encoder);
  }

  /**
   * Adds position data for an counter to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param counter Counter to add to DMA.
   */
  public void addCounter(Counter counter) {
    DMAJNI.addCounter(m_dmaHandle, counter.m_counter);
  }

  /**
   * Adds timer data for an counter to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param counter Counter to add to DMA.
   */
  public void addCounterPeriod(Counter counter) {
    DMAJNI.addCounterPeriod(m_dmaHandle, counter.m_counter);
  }

  /**
   * Adds a digital source to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param digitalSource DigitalSource to add to DMA.
   */
  public void addDigitalSource(DigitalSource digitalSource) {
    DMAJNI.addDigitalSource(m_dmaHandle, digitalSource.getPortHandleForRouting());
  }

  /**
   * Adds a duty cycle input to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param dutyCycle DutyCycle to add to DMA.
   */
  public void addDutyCycle(DutyCycle dutyCycle) {
    DMAJNI.addDutyCycle(m_dmaHandle, dutyCycle.m_handle);
  }

  /**
   * Adds an analog input to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param analogInput AnalogInput to add to DMA.
   */
  public void addAnalogInput(AnalogInput analogInput) {
    DMAJNI.addAnalogInput(m_dmaHandle, analogInput.m_port);
  }

  /**
   * Adds averaged data of an analog input to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param analogInput AnalogInput to add to DMA.
   */
  public void addAveragedAnalogInput(AnalogInput analogInput) {
    DMAJNI.addAveragedAnalogInput(m_dmaHandle, analogInput.m_port);
  }

  /**
   * Adds accumulator data of an analog input to be collected by DMA.
   *
   * <p>This can only be called if DMA is not started.
   *
   * @param analogInput AnalogInput to add to DMA.
   */
  public void addAnalogAccumulator(AnalogInput analogInput) {
    DMAJNI.addAnalogAccumulator(m_dmaHandle, analogInput.m_port);
  }

  /**
   * Sets an external DMA trigger.
   *
   * @param source the source to trigger from.
   * @param rising trigger on rising edge.
   * @param falling trigger on falling edge.
   * @return the index of the trigger
   */
  public int setExternalTrigger(DigitalSource source, boolean rising, boolean falling) {
    return DMAJNI.setExternalTrigger(
        m_dmaHandle,
        source.getPortHandleForRouting(),
        source.getAnalogTriggerTypeForRouting(),
        rising,
        falling);
  }

  /**
   * Sets a DMA PWM edge trigger.
   *
   * @param pwm the PWM to trigger from.
   * @param rising trigger on rising edge.
   * @param falling trigger on falling edge.
   * @return the index of the trigger
   */
  public int setPwmEdgeTrigger(PWM pwm, boolean rising, boolean falling) {
    return DMAJNI.setExternalTrigger(m_dmaHandle, pwm.getHandle(), 0, rising, falling);
  }

  /**
   * Sets a DMA PWMMotorController edge trigger.
   *
   * @param pwm the PWMMotorController to trigger from.
   * @param rising trigger on rising edge.
   * @param falling trigger on falling edge.
   * @return the index of the trigger
   */
  public int setPwmEdgeTrigger(PWMMotorController pwm, boolean rising, boolean falling) {
    return DMAJNI.setExternalTrigger(m_dmaHandle, pwm.getPwmHandle(), 0, rising, falling);
  }

  /**
   * Clear all sensors from the DMA collection list.
   *
   * <p>This can only be called if DMA is not started.
   */
  public void clearSensors() {
    DMAJNI.clearSensors(m_dmaHandle);
  }

  /**
   * Clear all external triggers from the DMA trigger list.
   *
   * <p>This can only be called if DMA is not started.
   */
  public void clearExternalTriggers() {
    DMAJNI.clearExternalTriggers(m_dmaHandle);
  }

  /**
   * Starts DMA Collection.
   *
   * @param queueDepth The number of objects to be able to queue.
   */
  public void start(int queueDepth) {
    DMAJNI.startDMA(m_dmaHandle, queueDepth);
  }

  /** Stops DMA Collection. */
  public void stop() {
    DMAJNI.stopDMA(m_dmaHandle);
  }
}

File: wpilibc/src/main/native/cpp/DMA.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/DMA.h"

#include <frc/AnalogInput.h>
#include <frc/Counter.h>
#include <frc/DigitalSource.h>
#include <frc/DutyCycle.h>
#include <frc/Encoder.h>
#include <frc/PWM.h>
#include <frc/motorcontrol/PWMMotorController.h>

#include <hal/DMA.h>
#include <hal/HALBase.h>

#include "frc/Errors.h"

using namespace frc;

DMA::DMA() {
  int32_t status = 0;
  dmaHandle = HAL_InitializeDMA(&status);
  FRC_CheckErrorStatus(status, "InitializeDMA");
}

void DMA::SetPause(bool pause) {
  int32_t status = 0;
  HAL_SetDMAPause(dmaHandle, pause, &status);
  FRC_CheckErrorStatus(status, "SetPause");
}

void DMA::SetTimedTrigger(units::second_t period) {
  int32_t status = 0;
  HAL_SetDMATimedTrigger(dmaHandle, period.value(), &status);
  FRC_CheckErrorStatus(status, "SetTimedTrigger");
}

void DMA::SetTimedTriggerCycles(int cycles) {
  int32_t status = 0;
  HAL_SetDMATimedTriggerCycles(dmaHandle, cycles, &status);
  FRC_CheckErrorStatus(status, "SetTimedTriggerCycles");
}

void DMA::AddEncoder(const Encoder* encoder) {
  int32_t status = 0;
  HAL_AddDMAEncoder(dmaHandle, encoder->m_encoder, &status);
  FRC_CheckErrorStatus(status, "AddEncoder");
}

void DMA::AddEncoderPeriod(const Encoder* encoder) {
  int32_t status = 0;
  HAL_AddDMAEncoderPeriod(dmaHandle, encoder->m_encoder, &status);
  FRC_CheckErrorStatus(status, "AddEncoderPeriod");
}

void DMA::AddCounter(const Counter* counter) {
  int32_t status = 0;
  HAL_AddDMACounter(dmaHandle, counter->m_counter, &status);
  FRC_CheckErrorStatus(status, "AddCounter");
}

void DMA::AddCounterPeriod(const Counter* counter) {
  int32_t status = 0;
  HAL_AddDMACounterPeriod(dmaHandle, counter->m_counter, &status);
  FRC_CheckErrorStatus(status, "AddCounterPeriod");
}

void DMA::AddDigitalSource(const DigitalSource* digitalSource) {
  int32_t status = 0;
  HAL_AddDMADigitalSource(dmaHandle, digitalSource->GetPortHandleForRouting(),
                          &status);
  FRC_CheckErrorStatus(status, "AddDigitalSource");
}

void DMA::AddDutyCycle(const DutyCycle* dutyCycle) {
  int32_t status = 0;
  HAL_AddDMADutyCycle(dmaHandle, dutyCycle->m_handle, &status);
  FRC_CheckErrorStatus(status, "AddDutyCycle");
}

void DMA::AddAnalogInput(const AnalogInput* analogInput) {
  int32_t status = 0;
  HAL_AddDMAAnalogInput(dmaHandle, analogInput->m_port, &status);
  FRC_CheckErrorStatus(status, "AddAnalogInput");
}

void DMA::AddAveragedAnalogInput(const AnalogInput* analogInput) {
  int32_t status = 0;
  HAL_AddDMAAveragedAnalogInput(dmaHandle, analogInput->m_port, &status);
  FRC_CheckErrorStatus(status, "AddAveragedAnalogInput");
}

void DMA::AddAnalogAccumulator(const AnalogInput* analogInput) {
  int32_t status = 0;
  HAL_AddDMAAnalogAccumulator(dmaHandle, analogInput->m_port, &status);
  FRC_CheckErrorStatus(status, "AddAnalogAccumulator");
}

int DMA::SetExternalTrigger(DigitalSource* source, bool rising, bool falling) {
  int32_t status = 0;
  int32_t idx =
      HAL_SetDMAExternalTrigger(dmaHandle, source->GetPortHandleForRouting(),
                                static_cast<HAL_AnalogTriggerType>(
                                    source->GetAnalogTriggerTypeForRouting()),
                                rising, falling, &status);
  FRC_CheckErrorStatus(status, "SetExternalTrigger");
  return idx;
}

int DMA::SetPwmEdgeTrigger(PWMMotorController* source, bool rising,
                           bool falling) {
  int32_t status = 0;
  int32_t idx = HAL_SetDMAExternalTrigger(
      dmaHandle, source->GetPwm()->m_handle,
      HAL_AnalogTriggerType::HAL_Trigger_kInWindow, rising, falling, &status);
  FRC_CheckErrorStatus(status, "SetPWmEdgeTrigger");
  return idx;
}

int DMA::SetPwmEdgeTrigger(PWM* source, bool rising, bool falling) {
  int32_t status = 0;
  int32_t idx = HAL_SetDMAExternalTrigger(
      dmaHandle, source->m_handle, HAL_AnalogTriggerType::HAL_Trigger_kInWindow,
      rising, falling, &status);
  FRC_CheckErrorStatus(status, "SetPWmEdgeTrigger");
  return idx;
}

void DMA::ClearSensors() {
  int32_t status = 0;
  HAL_ClearDMASensors(dmaHandle, &status);
  FRC_CheckErrorStatus(status, "ClearSensors");
}

void DMA::ClearExternalTriggers() {
  int32_t status = 0;
  HAL_ClearDMAExternalTriggers(dmaHandle, &status);
  FRC_CheckErrorStatus(status, "ClearExternalTriggers");
}

void DMA::Start(int queueDepth) {
  int32_t status = 0;
  HAL_StartDMA(dmaHandle, queueDepth, &status);
  FRC_CheckErrorStatus(status, "Start");
}

void DMA::Stop() {
  int32_t status = 0;
  HAL_StopDMA(dmaHandle, &status);
  FRC_CheckErrorStatus(status, "Stop");
}
